src/validation.cpp

```diff
@@ -2277,13 +2277,29 @@ bool CChainState::FlushStateToDisk(
         CoinsCacheSizeState cache_state = GetCoinsCacheSizeState();
         LOCK(m_blockman.cs_LastBlockFile);
         if (fPruneMode && (m_blockman.m_check_for_pruning || nManualPruneHeight > 0) && !fReindex) {
-            // make sure we don't prune above the blockfilterindexes bestblocks
+            // make sure we don't prune above any of the prune locks bestblocks
             // pruning is height-based
             int last_prune = m_chain.Height(); // last height we can prune
+            std::optional<std::string> limiting_lock; // prune lock that actually was the limiting factor, only used for logging
+
             ForEachBlockFilterIndex([&](BlockFilterIndex& index) {
-               last_prune = std::max(1, std::min(last_prune, index.GetSummary().best_block_height));
+                last_prune = std::max(1, std::min(last_prune, index.GetSummary().best_block_height));
             });
 
+            for (const auto& prune_lock : m_blockman.m_prune_locks) {
+                if (prune_lock.second.height_first == std::numeric_limits<int>::max()) continue;
```

Comment:
Yeah, at the current state of the code I don't see how this should happen. But I still would like to keep something there as belts & suspenders. But since this would be a developer error and not a usual case, I changed this line to an `assert` because it seems more appropriate here.